numba/core/types/abstract.py

```diff
@@ -1,12 +1,23 @@
 from abc import ABCMeta, abstractmethod, abstractproperty
-from typing import Dict as ptDict, Type as ptType
+import typing as pt
+
 import itertools
 import weakref
 
 import numpy as np
-
+from numba.core.typeconv import Conversion
+if pt.TYPE_CHECKING:
+    # these imports are only needed for python typing and would otherwise
+    # cause circular imports
+    from numba.core.typing.templates import Signature
+    from numba.core.types import Array
+    from numba.core.base import BaseContext
```

Comment:
Just to confirm, currently `Signature` and `BaseContext` are un-annotated (i.e. aliases for `pt.Any`)?  I had started annotating `Signature` on another branch.

Reply:
are they really aliases for `pt.Any` if they are not themselves annotated? 
In this example, mypy complains about `bar` but not about `baz`. Even if `Foo` is not annotated mypy correctly checks inputs annotated as `Foo`

```
from typing import Any
class Foo:
    def __init__(self, a):
        self.a = a

def bar(x: Foo) -> Foo:
    return x

def baz(x: Any) -> Any:
    return x

foo = Foo(1)
bar(1)
baz(1)
```

Reply:
I think I was incorrect for this case.  I have experience working with larger packages with more complex ways of defining public endpoints (e.g. Tensorflow).  I think for this case we do get some checking.  From the [mypy docs](https://mypy.readthedocs.io/en/stable/running_mypy.html):

> Even if mypy is unable to perfectly type check a file, it can still glean some useful information by parsing it (for example, understanding what methods a given object has).

Sorry for the confusion.